### PR TITLE
bugfix: deviceName garbled in Windows #34

### PR DESCRIPTION
**Pull request by nononoone**
Fix https://github.com/MadLittleMods/node-usb-detection/issues/34

Fix deviceName garbled in Windows

### CODE REVIEW

File: src/detection_win.cpp [hunk 1]
```diff
@@ -550,3 +551,22 @@ void UpdateDevice(PDEV_BROADCAST_DEVICEINTERFACE pDevInf, WPARAM wParam, DeviceS
...
+    std::wstring wstr_tmp(wstr_size, 0);
+    MultiByteToWideChar(CP_ACP, 0, str.c_str(), -1, &wstr_tmp[0], wstr_size);
+
+    //Wide character to Utf8
+    int str_size = WideCharToMultiByte(CP_UTF8, 0, &wstr_tmp[0], (int)wstr_tmp.size(), NULL, 0, NULL, NULL);
+    std::string str_utf8(str_size, 0);
+    WideCharToMultiByte(CP_UTF8, 0, &wstr_tmp[0], (int)wstr_tmp.size(), &str_utf8[0], str_size, NULL, NULL);
+
+    return str_utf8;
+}
```
Comment (MadLittleMods) [inline src/detection_win.cpp]: Did you get this code from somewhere like StackOverflow? Let's link it for reference like the following

```cpp
// via https://stackoverflow.com/q/25410284/796832
```

File: src/detection_win.cpp [hunk 2]
```diff
@@ -89,6 +89,7 @@ _SetupDiGetDeviceRegistryProperty DllSetupDiGetDeviceRegistryProperty;
  * Local Helper Functions protoypes
  **********************************/
 void UpdateDevice(PDEV_BROADCAST_DEVICEINTERFACE pDevInf, WPARAM wParam, DeviceState_t state);
+std::string Utf8Encode(const std::string &str);
```
Comment (MadLittleMods) [inline src/detection_win.cpp]: Is `std::string` still available after https://github.com/MadLittleMods/node-usb-detection/pull/84 is merged?

**Comment by MadLittleMods** [comment]
@nononoone Can you give a before after results when you use `usbDetect.find`

**Comment by nononoone** [comment]
> @nononoone Can you give a before after results when you use `usbDetect.find`?

before:
```
G:\Test\node_modules\usb-detection\test>node dev-helper-test.js
startMonitoring
find 14 [ { locationId: 0,
    vendorId: 0,
    productId: 0,
    deviceName: 'USB Root Hub',
    manufacturer: '(��׼ USB ��������)',
    serialNumber: '',
    deviceAddress: 12 },
  { locationId: 0,
    vendorId: 0,
    productId: 0,
    deviceName: 'USB Root Hub',
    manufacturer: '(��׼ USB ��������)',
    serialNumber: '',
    deviceAddress: 3 },
  { locationId: 0,
    vendorId: 0,
    productId: 0,
    deviceName: 'USB ��������(USB 3.0)',
    manufacturer: '(��׼ USB ������)',
    serialNumber: '',
    deviceAddress: 10 },
```
after:
```
G:\Working\Test\node_modules\usb-detection\test>node dev-helper-test.js  
startMonitoring
find 14 [ { locationId: 0,
    vendorId: 0,
    productId: 0,
    deviceName: 'USB Root Hub',
    manufacturer: '(标准 USB 主控制器)',
    serialNumber: '',
    deviceAddress: 12 },
  { locationId: 0,
    vendorId: 0,
    productId: 0,
    deviceName: 'USB Root Hub',
    manufacturer: '(标准 USB 主控制器)',
    serialNumber: '',
    deviceAddress: 3 },
  { locationId: 0,
    vendorId: 0,
    productId: 0,
    deviceName: 'USB 根集线器(USB 3.0)',
    manufacturer: '(标准 USB 集线器)',
    serialNumber: '',
    deviceAddress: 10 },
```

File: src/detection_win.cpp (hunk 2 above)
Comment (nononoone) [reply src/detection_win.cpp]: No problem, 'std::string' is the c++ standard，mean string with namespace std,
#84 problem is the 'ATL' is not the standard library.

File: src/detection_win.cpp (hunk 1 above)
Comment (nononoone) [reply src/detection_win.cpp]: No, i wrote it myself, it easy for me, i wrote is before. sometimes i need change the code page from 'gbk' to 'utf8'.

**Comment by TyQuanCN** [comment]
V4.3.0 encountered the same bug, I hope to be able to merge this code in the new version，please

File: src/detection_win.cpp (hunk 1 above)
Comment (MadLittleMods) [reply src/detection_win.cpp]: There is a bit too much magic here for me to be comfortable. Just seems like something could be exploitable here with all of this reference and byte copying stuff.

Would definitely prefer a standard approach to tackle this

File: src/detection_win.cpp [hunk 3]
```diff
@@ -550,3 +551,22 @@ void UpdateDevice(PDEV_BROADCAST_DEVICEINTERFACE pDevInf, WPARAM wParam, DeviceS
...
+std::string Utf8Encode(const std::string &str)
+{
+    if (str.empty()) {
+        return std::string();
+    }
+
+    //System default code page to wide character
+    int wstr_size = MultiByteToWideChar(CP_ACP, 0, str.c_str(), -1, NULL, 0);
+    std::wstring wstr_tmp(wstr_size, 0);
+    MultiByteToWideChar(CP_ACP, 0, str.c_str(), -1, &wstr_tmp[0], wstr_size);
```
Comment (MadLittleMods) [inline src/detection_win.cpp]: I see this is pretty similar to https://stackoverflow.com/a/6693107/796832 so seems good enough to me

File: src/detection_win.cpp (hunk 1 above)
Comment (MadLittleMods) [reply src/detection_win.cpp]: Perhaps https://stackoverflow.com/a/3999597/796832

File: src/detection_win.cpp [hunk 4]
```diff
@@ -550,3 +551,22 @@ void UpdateDevice(PDEV_BROADCAST_DEVICEINTERFACE pDevInf, WPARAM wParam, DeviceS
...
+
+    //Wide character to Utf8
+    int str_size = WideCharToMultiByte(CP_UTF8, 0, &wstr_tmp[0], (int)wstr_tmp.size(), NULL, 0, NULL, NULL);
+    std::string str_utf8(str_size, 0);
+    WideCharToMultiByte(CP_UTF8, 0, &wstr_tmp[0], (int)wstr_tmp.size(), &str_utf8[0], str_size, NULL, NULL);
```
Comment (MadLittleMods) [inline src/detection_win.cpp]: This seems pretty similar to WideCharToMultiByte so seems good enough to me

**Comment by MadLittleMods** [comment]
Sorry for the delay @nononoone 🥶

This is now shipped in [`usb-detection@4.4.0`](https://github.com/MadLittleMods/node-usb-detection/blob/master/CHANGELOG.md#440---2019-10-6) :tada:

Thanks for the fix ❤️

**Pull request merged**